File: backend/app/schemas/orchestration.py
"""Schemas for Phase 6 subscriptions, charts, and organize boundaries."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, model_validator

from .acquisition import PathHandoffInfo, SearchCandidateDetail, SearchJobSummary
from .integration import AdapterMode, AdapterResolution, VerificationState
from .metadata import MetadataDetail
from .music_media import (
    MusicMediaInfo,
    MusicMediaInput,
    MusicMetaBase,
    MusicRecognitionAssessment,
    MusicRecognitionState,
)
from .mvp import EntityType


class SubscriptionType(str, Enum):
    ARTIST = "artist"
    ALBUM = "album"
    TRACK = "track"
    CHART_ENTRY = "chart_entry"


class SubscriptionState(str, Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    ARCHIVED = "archived"


class SubscriptionMode(str, Enum):
    MANUAL = "manual"
    SCHEDULED = "scheduled"
    SCHEDULED_PLACEHOLDER = "scheduled_placeholder"


class SubscriptionRunStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    MATCHED = "matched"
    DISPATCHED = "dispatched"
    APPLIED = "applied"
    MANUAL_PENDING = "manual_pending"
    NO_RESULT = "no_result"
    FAILED = "failed"


class OrganizeStatus(str, Enum):
    PLANNED = "planned"
    PREVIEW_READY = "preview_ready"
    APPLY_PENDING = "apply_pending"
    APPLIED = "applied"
    FALLBACK_APPLIED = "fallback_applied"
    SKIPPED = "skipped"
    FAILED = "failed"


class OrganizeConflictPolicy(str, Enum):
    SKIP_EXISTING = "skip_existing"
    OVERWRITE = "overwrite"
    APPEND_SUFFIX = "append_suffix"


class ChartProviderInfo(BaseModel):
    id: str
    chart_source: str
    display_name: str
    enabled: bool = True
    mock: bool = True
    note: str
    integration_point: str


class ChartInfo(BaseModel):
    id: str
    chart_source: str
    chart_name: str
    chart_type: EntityType
    region: str | None = None
    category: str | None = None
    refresh_hint: str | None = None
    item_count: int = 0
    updated_at: datetime
    mock: bool = True
    note: str
    summary: str | None = None
    chart_group: str | None = None
    chart_scope: str | None = None
    freshness_label: str | None = None
    supports_subscription: bool = True


class ChartEntryInfo(BaseModel):
    item_id: str
    chart_id: str
    chart_source: str
    chart_name: str
    rank: int
    item_type: EntityType
    target_id: str
    target_name: str
    subtitle: str | None = None
    provider: str
    source_type: str
    target_payload: dict[str, Any] = Field(default_factory=dict)
    mock: bool = True
    note: str


class DiscoveryEntryView(BaseModel):
    entry: ChartEntryInfo
    media_input: MusicMediaInput
    meta_base: MusicMetaBase
    entry_summary: str
    badges: list[str] = Field(default_factory=list)
    highlight_reason: str | None = None
    recognition_assessment: MusicRecognitionAssessment = Field(
        default_factory=lambda: MusicRecognitionAssessment(state=MusicRecognitionState.READY)
    )


class DiscoveryEntryGroup(BaseModel):
    group_key: str
    group_label: str
    items: list[DiscoveryEntryView] = Field(default_factory=list)


class ChartListData(BaseModel):
    items: list[ChartInfo] = Field(default_factory=list)
    total: int = 0
    mock: bool = True
    note: str
    integration_point: str


class ChartDetailData(BaseModel):
    chart: ChartInfo
    items: list[ChartEntryInfo] = Field(default_factory=list)
    item_count: int = 0
    mock: bool = True
    note: str
    integration_point: str
    hero_entry: DiscoveryEntryView | None = None
    summary_stats: dict[str, str | int] = Field(default_factory=dict)
    entry_groups: list[DiscoveryEntryGroup] = Field(default_factory=list)
    recognition_summary: dict[str, int] = Field(default_factory=dict)


class CreateChartEntrySubscriptionRequest(BaseModel):
    chart_item_id: str
    mode: SubscriptionMode = SubscriptionMode.MANUAL
    preference_json: dict[str, Any] = Field(default_factory=dict)


class CreateSubscriptionRequest(BaseModel):
    subscription_type: SubscriptionType
    target_id: str
    target_name: str | None = None
    target_entity_type: EntityType | None = None
    mode: SubscriptionMode = SubscriptionMode.MANUAL
    preference_json: dict[str, Any] = Field(default_factory=dict)
    target_payload: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def validate_target(self) -> "CreateSubscriptionRequest":
        if self.subscription_type == SubscriptionType.CHART_ENTRY and self.target_entity_type is None:
            raise ValueError("chart_entry subscription requires target_entity_type.")
        return self


class UpdateSubscriptionRequest(BaseModel):
    status: SubscriptionState | None = None
    mode: SubscriptionMode | None = None
    preference_json: dict[str, Any] | None = None


class SubscriptionSummary(BaseModel):
    id: str
    subscription_type: SubscriptionType
    target_id: str
    target_name: str
    target_entity_type: EntityType | None = None
    chart_source: str | None = None
    chart_name: str | None = None
    status: SubscriptionState
    mode: SubscriptionMode
    preference_json: dict[str, Any] = Field(default_factory=dict)
    target_payload: dict[str, Any] = Field(default_factory=dict)
    music_media_input: MusicMediaInput | None = None
    music_meta_base: MusicMetaBase | None = None
    music_recognition_assessment: MusicRecognitionAssessment | None = None
    music_media_info: MusicMediaInfo | None = None
    latest_run_status: str | None = None
    last_run_at: datetime | None = None
    mock: bool = True
    note: str | None = None
    created_at: datetime
    updated_at: datetime


class OrganizeStrategySnapshot(BaseModel):
    strategy_name: str
    library_type: str = "music"
    root_path: str
    artist_dir_template: str
    album_dir_template: str
    track_file_template: str
    conflict_policy: OrganizeConflictPolicy = OrganizeConflictPolicy.SKIP_EXISTING
    template_note: str


class OrganizePlan(BaseModel):
    strategy: str
    strategy_snapshot: OrganizeStrategySnapshot
    target_library_path: str
    target_relative_path: str
    strategy_note: str


class OrganizePreviewResult(BaseModel):
    id: str
    subscription_run_id: str | None = None
    search_job_id: str | None = None
    candidate_id: str | None = None
    binding_id: str | None = None
    organizeable: bool
    organize_backend: AdapterMode = AdapterMode.MOCK
    adapter_mode: AdapterMode = AdapterMode.MOCK
    strategy: str = "music_default_layout"
    strategy_snapshot: OrganizeStrategySnapshot
    organize_status: OrganizeStatus
    target_library_path: str
    target_relative_path: str
    strategy_note: str
    integration_point: str
    capability_source: str = "mock.adapter"
    fallback_reason: str | None = None
    failure_reason: str | None = None
    path_handoff: PathHandoffInfo | None = None
    verification_state: VerificationState = VerificationState.PLACEHOLDER
    adapter_resolution: AdapterResolution | None = None
    music_media_input: MusicMediaInput | None = None
    music_meta_base: MusicMetaBase | None = None
    music_recognition_assessment: MusicRecognitionAssessment | None = None
    music_media_info: MusicMediaInfo | None = None
    mock: bool = True
    note: str | None = None
    created_at: datetime
    updated_at: datetime


class SubscriptionRunSummary(BaseModel):
    id: str
    subscription_id: str
    search_job_id: str | None = None
    execution_status: SubscriptionRunStatus
    matched_candidates_count: int
    organize_record_id: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    summary_json: dict[str, Any] = Field(default_factory=dict)
    music_media_input: MusicMediaInput | None = None
    music_meta_base: MusicMetaBase | None = None
    music_recognition_assessment: MusicRecognitionAssessment | None = None
    music_media_info: MusicMediaInfo | None = None
    mock: bool = True
    note: str | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime


class SubscriptionDetail(SubscriptionSummary):
    recent_runs: list[SubscriptionRunSummary] = Field(default_factory=list)


class SubscriptionListData(BaseModel):
    items: list[SubscriptionSummary] = Field(default_factory=list)
    total: int = 0
    mock: bool = True
    note: str


class SubscriptionRunListData(BaseModel):
    subscription_id: str
    items: list[SubscriptionRunSummary] = Field(default_factory=list)
    total: int = 0
    mock: bool = True
    note: str


class SubscriptionRunDetail(SubscriptionRunSummary):
    subscription: SubscriptionSummary
    metadata_target: MetadataDetail | None = None
    search_job: SearchJobSummary | None = None
    candidates: list[SearchCandidateDetail] = Field(default_factory=list)
    organize_preview: OrganizePreviewResult | None = None


class OrganizePreviewRequest(BaseModel):
    candidate_id: str | None = None
    binding_id: str | None = None

    @model_validator(mode="after")
    def validate_target(self) -> "OrganizePreviewRequest":
        if not self.candidate_id and not self.binding_id:
            raise ValueError("candidate_id or binding_id is required.")
        return self


class OrganizeApplyRequest(BaseModel):
    organize_job_id: str


class OrganizeRecordListData(BaseModel):
    items: list[OrganizePreviewResult] = Field(default_factory=list)
    total: int = 0
    mock: bool = True
    note: str


class OrganizeAdapterResult(BaseModel):
    organizeable: bool
    organize_backend: AdapterMode = AdapterMode.MOCK
    adapter_mode: AdapterMode = AdapterMode.MOCK
    strategy: str = "music_default_layout"
    strategy_snapshot: OrganizeStrategySnapshot
    organize_status: OrganizeStatus
    target_library_path: str
    target_relative_path: str
    strategy_note: str
    integration_point: str
    capability_source: str = "mock.adapter"
    fallback_reason: str | None = None
    failure_reason: str | None = None
    path_handoff: PathHandoffInfo | None = None
    verification_state: VerificationState = VerificationState.PLACEHOLDER
    adapter_resolution: AdapterResolution | None = None
    mock: bool = True
    note: str

File: frontend/src/types/orchestration.ts
import type {
  AdapterMode,
  AdapterResolution,
  PathHandoffInfo,
  SearchCandidateDetail,
  SearchJobSummary,
  VerificationState,
} from '@/types/acquisition';
import type { ApiResponse, EntityType, MetadataDetail } from '@/types/metadata';

export type SubscriptionType = 'artist' | 'album' | 'track' | 'chart_entry';
export type SubscriptionState = 'active' | 'paused' | 'archived';
export type SubscriptionMode = 'manual' | 'scheduled_placeholder';
export type SubscriptionRunStatus =
  | 'queued'
  | 'running'
  | 'matched'
  | 'manual_pending'
  | 'no_result'
  | 'failed';
export type OrganizeStatus =
  | 'planned'
  | 'preview_ready'
  | 'apply_pending'
  | 'applied'
  | 'fallback_applied'
  | 'skipped'
  | 'failed';
export type OrganizeConflictPolicy = 'skip_existing' | 'overwrite' | 'append_suffix';

export interface ChartProviderInfo {
  id: string;
  chart_source: string;
  display_name: string;
  enabled: boolean;
  mock: boolean;
  note: string;
  integration_point: string;
}

export interface ChartInfo {
  id: string;
  chart_source: string;
  chart_name: string;
  chart_type: EntityType;
  region?: string | null;
  category?: string | null;
  refresh_hint?: string | null;
  item_count: number;
  updated_at: string;
  mock: boolean;
  note: string;
}

export interface ChartEntryInfo {
  item_id: string;
  chart_id: string;
  chart_source: string;
  chart_name: string;
  rank: number;
  item_type: EntityType;
  target_id: string;
  target_name: string;
  subtitle?: string | null;
  provider: string;
  source_type: string;
  mock: boolean;
  note: string;
}

export interface ChartListData {
  items: ChartInfo[];
  total: number;
  mock: boolean;
  note: string;
  integration_point: string;
}

export interface ChartDetailData {
  chart: ChartInfo;
  items: ChartEntryInfo[];
  item_count: number;
  mock: boolean;
  note: string;
  integration_point: string;
}

export interface CreateChartEntrySubscriptionPayload {
  chart_item_id: string;
  mode?: SubscriptionMode;
  preference_json?: Record<string, unknown>;
}

export interface CreateSubscriptionPayload {
  subscription_type: SubscriptionType;
  target_id: string;
  target_name?: string;
  target_entity_type?: EntityType;
  mode?: SubscriptionMode;
  preference_json?: Record<string, unknown>;
  target_payload?: Record<string, unknown>;
}

export interface UpdateSubscriptionPayload {
  status?: SubscriptionState;
  mode?: SubscriptionMode;
  preference_json?: Record<string, unknown> | null;
}

export interface SubscriptionSummary {
  id: string;
  subscription_type: SubscriptionType;
  target_id: string;
  target_name: string;
  target_entity_type?: EntityType | null;
  chart_source?: string | null;
  chart_name?: string | null;
  status: SubscriptionState;
  mode: SubscriptionMode;
  preference_json: Record<string, unknown>;
  target_payload: Record<string, unknown>;
  latest_run_status?: string | null;
  last_run_at?: string | null;
  mock: boolean;
  note?: string | null;
  created_at: string;
  updated_at: string;
}

export interface OrganizeStrategySnapshot {
  strategy_name: string;
  library_type: string;
  root_path: string;
  artist_dir_template: string;
  album_dir_template: string;
  track_file_template: string;
  conflict_policy: OrganizeConflictPolicy;
  template_note: string;
}

export interface OrganizePreviewResult {
  id: string;
  subscription_run_id?: string | null;
  search_job_id?: string | null;
  candidate_id?: string | null;
  binding_id?: string | null;
  organizeable: boolean;
  organize_backend: AdapterMode;
  adapter_mode: AdapterMode;
  strategy: string;
  strategy_snapshot: OrganizeStrategySnapshot;
  organize_status: OrganizeStatus;
  target_library_path: string;
  target_relative_path: string;
  strategy_note: string;
  integration_point: string;
  capability_source: string;
  fallback_reason?: string | null;
  failure_reason?: string | null;
  path_handoff?: PathHandoffInfo | null;
  verification_state: VerificationState;
  adapter_resolution?: AdapterResolution | null;
  mock: boolean;
  note?: string | null;
  created_at: string;
  updated_at: string;
}

export interface SubscriptionRunSummary {
  id: string;
  subscription_id: string;
  search_job_id?: string | null;
  execution_status: SubscriptionRunStatus;
  matched_candidates_count: number;
  organize_record_id?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  summary_json: Record<string, unknown>;
  mock: boolean;
  note?: string | null;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
}

export interface SubscriptionDetail extends SubscriptionSummary {
  recent_runs: SubscriptionRunSummary[];
}

export interface SubscriptionListData {
  items: SubscriptionSummary[];
  total: number;
  mock: boolean;
  note: string;
}

export interface SubscriptionRunListData {
  subscription_id: string;
  items: SubscriptionRunSummary[];
  total: number;
  mock: boolean;
  note: string;
}

export interface SubscriptionRunDetail extends SubscriptionRunSummary {
  subscription: SubscriptionSummary;
  metadata_target?: MetadataDetail | null;
  search_job?: SearchJobSummary | null;
  candidates: SearchCandidateDetail[];
  organize_preview?: OrganizePreviewResult | null;
}

export interface OrganizePreviewPayload {
  candidate_id?: string;
  binding_id?: string;
}

export interface OrganizeApplyPayload {
  organize_job_id: string;
}

export interface OrganizeRecordListData {
  items: OrganizePreviewResult[];
  total: number;
  mock: boolean;
  note: string;
}

export type ApiChartProvidersResponse = ApiResponse<ChartProviderInfo[]>;
export type ApiChartListResponse = ApiResponse<ChartListData>;
export type ApiChartDetailResponse = ApiResponse<ChartDetailData>;
export type ApiSubscriptionResponse = ApiResponse<SubscriptionSummary>;
export type ApiSubscriptionDetailResponse = ApiResponse<SubscriptionDetail>;
export type ApiSubscriptionListResponse = ApiResponse<SubscriptionListData>;
export type ApiSubscriptionRunResponse = ApiResponse<SubscriptionRunDetail>;
export type ApiSubscriptionRunsResponse = ApiResponse<SubscriptionRunListData>;
export type ApiOrganizePreviewResponse = ApiResponse<OrganizePreviewResult>;
export type ApiOrganizeRecordResponse = ApiResponse<OrganizePreviewResult>;
export type ApiOrganizeJobsResponse = ApiResponse<OrganizeRecordListData>;
